routing/src/interface.rs

```diff
@@ -0,0 +1,363 @@
+    pub fn set_iftype(&mut self, iftype: IfType) {
+        self.iftype = iftype;
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Set the description of an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn set_description(&mut self, description: &str) {
+        self.description = Some(description.to_owned());
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Set the operational state of an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn set_oper_state(&mut self, state: IfState) {
+        if self.oper_state != state {
+            self.oper_state = state;
+            // Todo: log change
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Set the administrative state of an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn set_admin_state(&mut self, state: IfState) {
+        if self.admin_state != state {
+            self.admin_state = state;
+            // Todo: log change
+        }
+    }
+    // Todo: decide if set_oper_state() and set_admin_state() will be
+    // the ones to trigger the enforcement of the new state
+    // (e.g. by requesting the interface manager to bring up/down
+    // interfaces in kernel), or, instead, they just update the
+    // interface object seen by the rest of the routing system.
+
+    //////////////////////////////////////////////////////////////////
+    /// Attach an interface to a VRF. It is assumed that both the
+    /// interface and the VRF exist.
+    //////////////////////////////////////////////////////////////////
+    pub fn attach(&mut self, vrf: &Arc<Vrf>) -> Result<(), RouterError> {
+        if let Some(exist_vrf) = &self.vrf {
+            if !Arc::ptr_eq(exist_vrf, vrf) {
+                Err(RouterError::AlreadyAttached)
+            } else {
+                Ok(())
+            }
+        } else {
+            self.vrf = Some(vrf.clone());
+            Ok(())
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Detach an interface from its VRF (whichever it is)
+    //////////////////////////////////////////////////////////////////
+    pub fn detach(&mut self) {
+        self.vrf.take();
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Get the VRF that an interface is attached to, or None otherwise
+    //////////////////////////////////////////////////////////////////
+    pub fn get_vrf(&self) -> Option<&Vrf> {
+        self.vrf.as_deref()
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Add (assign) an IP address to an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn add_ifaddr(&mut self, ifaddr: &IfAddress) {
+        self.addresses.insert(*ifaddr);
```

Comment:
very much not blocking, but I do suggest making this function return `bool` to indicate if the insert did anything.

Also, since `IfAddress: Copy` you don't need to pass by reference here.  You can just send in a `IfAddress` I think

Reply:
> very much not blocking, but I do suggest making this function return bool to indicate if the insert did anything.

Why? Note that it's a HashSet. If the (address,len) existed, it will be replaced (i.e noop). If it  didn't, it will surely be added. So, what would knowing if it existed or not help the caller?